add pgn130576 Small Craft Status / Trim Tab Position

nagro2 opened the pull request:
Add PGN130756 Small Craft Status / Trim Tab Position. 
Nicholas Agro 1/28/2018

Testing and validation performed:
- Successfully used this PGN to read and display Trim Tab Position from Garmin GBT 10 Trim Tab Position sender unit
- Successfully read Trim Tab Position messages output from this PGN on Actisense NMEA reader
- Succesfully displayed Trim Tab Position from this PGN on Garmin 4212 Chartplotter display

ttlappalainen commented:
Thank. Some fixes:

You should change parameters to int8_t to inform that they are signed:
void SetN2kPGN130576(tN2kMsg &N2kMsg, int8_t PortTrimTab, int8_t StbdTrimTab);
This is since for unavailable value you have to use N2kInt8NA. You can test is some field signed or not with NMEA Reader. For e.g. byte field provide 0x7f or 0xff. If NMEA reader shows -1 for 0xff, then value is signed and should show NA for 0x7f.

To save calls on padding you can use Add2ByteUInt(0xffff); Add4ByteUInt(0xffffffff). The padding is not required. Only the reserved bits has to be filled with 1. So if you know that reserved field 3 is just 8 bits, it is enough just have one AddByte(0xff). If you know that it is 16 bits, then just use Add2ByteUInt(0xffff); If you do not know, then fill whole message. You can check required padding from certified device message length. If length is 3, you need 1 byte padding, since 2 bytes carries data etc. It is important that the length is right and matches to message documentation. Othwerwise some devices may reject mesage.

ttlappalainen commented:
Also please update README.adoc with changes description and library.json and library.properties by incrementing last value. In principle this is new feature, but since it is not in core module, we just increment last value.

nagro2 commented:
Thanks very much for your feedback. Your library is awesome. I had been frustrated for years
at the closed nature of nmea2000 and was very happy to find your library. I consider it a privilege to be able to contribute.

You were correct about the trim tab values being int8_t. As you stated, when I set the value
to 0xFF, Actisense NMEA reader showed -1 and when I set it to 0x7F, NMEA reader showed NA. I have changed the trim tab position type to int8_t.

Padding is a bit more difficult. I could not find documentation to show certified device message
length. NMEA reader listed three fields, port, starboard and reserved, but did not give length of the reserved field. Originally I assumed the reserved field was 8 bits and configured it this way. However although NMEA reader showed a 3 byte payload for the messages from my PGN130576, it showed 6 0xFF bytes in the PGN130576 messages from the Garmin GBT10 trim tab position sensor. When I pad my PGN130576 definition to have 6 0xFF bytes it looks identical to the Garmin messages. I suppose there is no guarantee that Garmin is implementing the standard properly but at least it is an example from working hardware. I changed the 6 byte reserved field to the format you recommended.

I tested the library as before, by sending the PGN and confirming it on NMEA reader, and also receiving and correctly interpreting this PGN from the Garmin GBT10.

I updated README.adoc to reflect my addition. 

I was not sure how to update library.json and library.properties. I researched but did not see detailed explanations. I added my name in the author field but I don't know if that's correct. I would have expected to add my name to a field called contributor rather than author but there is no contributor field. Please let me know if I should do something else.

Thanks again.

ttlappalainen commented:
If NMEA Reader shows 3 byte payload for Garmin, we could expect that it is right and messages has only 1 byte reserved field and so we should pad only one reserved byte with 0xff. The result may not look identical, since on CAN traffic some systems pads rest of frame with 0xFF and some does not. It is more important that the length of message is right and so should be 3. Note that GBT10 seem to be on list of certified devices, so we should rely that. 

Currently I have seen requirement for padding only for last frame of fast packet and that is implemented. The possible padding for less that 8 byte single messages should be done on "CAN driver" level so e.g. on FlexCAN library.

Please remove your name from  library.json and library.properties authors. So only last number on version should be updated. Instead set on ReadMe.adoc:
Added: PGN130576 Small Craft Status / Trim Tab Position definition by Nicholas Agro
Otherwise I should add several others to authors. And I see that authors on those files points to persons, who actively develop the library.

nagro2 commented:
I changed the padding on pgn130576 to one 0xFF reserved byte per your recommendation and made the corrections to readme.adoc, library.json, and library.properties. Thanks again.

virtuvas commented:
hello all,

just finished interfacing my Bennett trim tab indicators with a teensy 3.5 and sent the PGN130576 to the bus. All fine BUT
Garmin GMI10 works fine, shows right %, pops up the tabs screen if you're on another display, all's fine.
Maretron DSM250 is not impressed, does nothing.

Now, I do have the original Bennett NMEA2000 adapter (a Lowrance hardware) that came with the system. It's crap when you hook it up, both the Garmin and the Maretron flash from 0 to 100% and loop that all the time, useless thing.
So what I did is looked at what the sentence I built using your lib shows and how it compares to the Lowrance one (that the Maretron also likes)

Turns out that there's only one difference, the padding at the end of the sentence:

my sentence is: 0A 7F FF (that is 10%) with Priority = 2, Length =3
Lowrance is: 00 00 00 FF FF FF FF FF (that is both 0%) with Priority = 2, BUT Length =8

Bearing in mind that Timo your sentence (with Length of 3bytes) only works on Garmin whereas Lowerance sentence with Length of 8bytes) works on both it is safe to assume that we should pad that to reach 8bytes.

Now, if I knew how to do it, I'd do and test it on the spot, so please your thoughts (and hopefully) editing on the lib so that I can test that as well now that I have both devices in my office playing!

cheers

V.

virtuvas commented:
doh, wasn't that difficult after looking at the N2kMessages.cpp

changed it to:
`
//*****************************************************************************
// Trim Tab Position
// Trim tab position is a percentage 0 to 100% where 0 is fully retracted and 100 is fully extended
void SetN2kPGN130576(tN2kMsg &N2kMsg, int8_t PortTrimTab, int8_t StbdTrimTab) {
    N2kMsg.SetPGN(130576L);
    N2kMsg.Priority=2;
    N2kMsg.AddByte(PortTrimTab);
    N2kMsg.AddByte(StbdTrimTab);
    N2kMsg.AddByte(0x00);;// Reserved.
    N2kMsg.AddByte(0xFF);;// Reserved.
    N2kMsg.AddByte(0xFF);;// Reserved.
    N2kMsg.AddByte(0xFF);;// Reserved.
    N2kMsg.AddByte(0xFF);;// Reserved.
    N2kMsg.AddByte(0xFF);;// Reserved.

}
`

and it now works on both Garmin and Maretron.
could you please edit your code to the same effect, I guess I shouldn't add 5 times an AddByte, must be a better way of doing it

cheers

V.

ttlappalainen commented:
Often on short messages length is 8 and rest is padded with 0xff. Some devices seem to be strict with length.

I do not understand 0 byte after Port/Stdb data. It should be reserved and 0xff instead of 0. Please try
```
    N2kMsg.AddByte(PortTrimTab);
    N2kMsg.AddByte(StbdTrimTab);
    N2kMsg.Add2ByteUInt(0xFFFF); // Reserved.
    N2kMsg.Add4ByteUInt(0xFFFFFFFF); // Reserved.

```
I'll fix that to the library.

virtuvas commented:
Timo,

you're right as always, works fine with your code, although it is slightly different on that byte to the Lowrance one.
So please fix that in the lib.

Now, if we could only find someone with the Maretron TMP100 box and get the custom EGT PGN130823 would be very nice!
